test/integration/http2_flood_integration_test.cc

```diff
@@ -1162,4 +1162,82 @@ TEST_P(Http2FloodMitigationTest, UpstreamEmptyHeaders) {
             test_server_->counter("cluster.cluster_0.http2.inbound_empty_frames_flood")->value());
 }
 
+// Verify that the HTTP/2 connection is terminated upon receiving invalid HEADERS frame.
+TEST_P(Http2FloodMitigationTest, UpstreamZerolenHeader) {
+  if (!initializeUpstreamFloodTest()) {
+    return;
+  }
+
+  // Send client request which will send an upstream request.
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto response = codec_client_->makeHeaderOnlyRequest(default_request_headers_);
+  waitForNextUpstreamRequest();
+
+  // Send an upstream reply.
+  auto* upstream = fake_upstreams_.front().get();
+  auto buf = Http2Frame::makeMalformedRequestWithZerolenHeader(Http2Frame::makeClientStreamId(0),
```

Comment:
I see. There is no counter specific to 0 length headers and both presence of invalid headers in response (i.e. path) and 0 length header increment the same `rx_messaging_error` counter.
The test is almost correct. It fails for a different reason though. The `makeMalformedRequestWithZerolenHeader` method will create invalid response header because it does not have `status` and has request headers like `method`, etc that are invalid in responses. So the nghttp2 will fail before it gets to the 0 len header. To make this test work you can add new `Http2Frame::makeMalformedResponsetWithZerolenHeader` method, that adds status header and 0 len header. See the `Http2Frame::makeMalformedRequest` on how to add status header.

Reply:
Yes, you are correct.
I've made the changes you've requested and now see the codec error due to the 0-length header:
`invalid http2: Invalid HTTP header field was received: frame type: 1, stream: 1, name: [], value: []`
In the log.
Not sure how to verify this in the test though, as you've mentioned there's only a generic counter.